nextstrain/cli/runner/managed_conda.py

```diff
@@ -0,0 +1,363 @@
+            f"unknown content-type for micromamba dist: {content_type}"
+
+        with tarfile.open(fileobj = response.raw, mode = "r|*") as tar:
+            # Ignore archive members starting with "/" and or including ".." parts,
+            # as these can be used (maliciously or accidentally) to overwrite
+            # unintended files (e.g. files outside of MICROMAMBA_ROOT).
+            safe_members = (
+                member
+                    for member in tar
+                     if not member.name.startswith("/")
+                    and ".." not in PurePosixPath(member.name).parts)
+
+            print(f"Downloading and extracting Micromamba to {MICROMAMBA_ROOT}…")
+            tar.extractall(path = str(MICROMAMBA_ROOT), members = safe_members)
+    else:
+        print(f"Downloading and extracting Micromamba to {MICROMAMBA_ROOT}…")
+
+    return True
+
+
+def setup_prefix(dry_run: bool = False, force: bool = False) -> bool:
+    """
+    Install Conda packages with Micromamba into our ``PREFIX``.
+    """
+    if not force and (PREFIX_BIN / "augur").exists():
+        print(f"Using existing Conda packages in {PREFIX}.")
+        print(f"  Hint: if you want to ignore this existing installation, re-run `nextstrain setup` with --force.")
+        return True
+
+    if PREFIX.exists():
+        print(f"Removing existing directory {PREFIX} to start fresh…")
+        if not dry_run:
+            shutil.rmtree(str(PREFIX))
+
+    # Conda packages to install, based on our unmanaged "native" install docs.
+    #
+    # Includes nextstrain-cli even though that's us and we're already installed
+    # and running because our own executable may not be on PATH in the runtime
+    # environment.
+    #
+    # Adds bash so that a newer version is guaranteed to be available on
+    # systems with an older bash.  This is similar to how our Docker runtime
+    # image includes its own bash too.
+    #
+    # HEY YOU: If you add/remove packages here, make sure to account for how
+    # update() should make the same changes to existing envs.
+    #   -trs, 1 Sept 2022
+    packages = (
+        "augur",
+        "auspice",
+        "nextalign",
+        "nextclade",
+        "nextstrain-cli",
+
+        "bash",
+        "epiweeks",
+        "git",
+        "pangolearn",
+        "pangolin",
+        "snakemake",
+    )
+
+    # Create environment
+    create = micromamba(
+        "create",
+
+        # Path-based env
+        "--prefix", PREFIX,
+
+        # BioConda config per <https://bioconda.github.io/#usage>
+        "--override-channels",
+        "--strict-channel-priority",
+        "--channel", "conda-forge",
+        "--channel", "bioconda",
+        "--channel", "defaults",
+
+        *packages,
+    )
+
+    print(f"Installing Conda packages into {PREFIX}…")
+    for pkg in packages:
+        print(f"  - {pkg}")
+
+    if not dry_run:
+        try:
+            subprocess.run(create, check = True)
```

Comment:
I found an edge case error while running `nextstrain setup managed-conda`, reproducible on both macOS and Linux:

1. Start fresh: `rm -r ~/.nextstrain/runtimes`.
2. Run `nextstrain setup managed-conda`.
3. Send keyboard interrupt after the `Downloading and extracting Micromamba ...` message.
4. Run `nextstrain setup managed-conda` again.
5. Observe `PermissionError`.

Running `rm -r ~/.nextstrain/runtimes` again fixes it.

<details>
<summary>Detailed output</summary>

```
$ nextstrain setup managed-conda
Setting up managed-conda…
Requesting Micromamba from https://api.anaconda.org/download/conda-forge/micromamba/0.25.1/linux-64/micromamba-0.25.1-0.tar.bz2…
Downloading and extracting Micromamba to /home/codespace/.nextstrain/runtimes/managed-conda/micromamba…
^CTraceback (most recent call last):
  File "runpy", line 196, in _run_module_as_main
  File "runpy", line 86, in _run_code
  File "nextstrain.cli.__main__", line 55, in <module>
  File "nextstrain.cli.__main__", line 19, in main
  File "nextstrain.cli", line 33, in run
  File "nextstrain.cli.console", line 36, in decorated
  File "nextstrain.cli.command.setup", line 64, in run
  File "nextstrain.cli.runner.managed_conda", line 76, in setup
  File "nextstrain.cli.runner.managed_conda", line 148, in setup_micromamba
  File "tarfile", line 2049, in extractall
  File "tarfile", line 2090, in extract
  File "tarfile", line 2163, in _extract_member
  File "tarfile", line 2212, in makefile
  File "tarfile", line 248, in copyfileobj
  File "tarfile", line 522, in read
  File "tarfile", line 544, in _read
KeyboardInterrupt

$ nextstrain setup managed-conda
Setting up managed-conda…
Using existing Micromamba installation at /home/codespace/.nextstrain/runtimes/managed-conda/micromamba.
  Hint: if you want to ignore this existing installation, re-run `nextstrain setup` with --force.
Installing Conda packages into /home/codespace/.nextstrain/runtimes/managed-conda/env…
  - augur
  - auspice
  - nextalign
  - nextclade
  - nextstrain-cli
  - bash
  - epiweeks
  - git
  - pangolearn
  - pangolin
  - snakemake
Error running ('/home/codespace/.nextstrain/runtimes/managed-conda/micromamba/bin/micromamba', '--root-prefix', '/home/codespace/.nextstrain/runtimes/managed-conda/micromamba', '--no-rc', '--no-env', '--yes', 'create', '--prefix', '/home/codespace/.nextstrain/runtimes/managed-conda/env', '--override-channels', '--strict-channel-priority', '--channel', 'conda-forge', '--channel', 'bioconda', '--channel', 'defaults', 'augur', 'auspice', 'nextalign', 'nextclade', 'nextstrain-cli', 'bash', 'epiweeks', 'git', 'pangolearn', 'pangolin', 'snakemake')
Traceback (most recent call last):
  File "nextstrain.cli.runner.managed_conda", line 220, in setup_prefix
    subprocess.run(create, check = True)
  File "subprocess", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "subprocess", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "subprocess", line 1842, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: '/home/codespace/.nextstrain/runtimes/managed-conda/micromamba/bin/micromamba'

Set up failed!
```

</details>

Reply:
This is likely because you've interrupted the tar archive extraction between when it writes the files to disk and when it restores modes like the executable bit (which it does at the very end). So the `micromamba` executable is there but not marked executable, and thus trying to execute it gets the OS error above.

Running `nextstrain setup --force managed-conda` would ignore the existing setup and start fresh without you having to manually `rm` the partial setup. I'm not sure it's worth playing whack-a-mole to try to cover the many ways partial setups can break assumptions and fail. I think instead I'd rather keep the simplifying assumptions and either a) deal with "fixing" broken setups by just redoing them fresh or b) avoid whack-a-mole by implementing transactions for setup (e.g. either it all completes successfully or it's all removed on failure). I think (a) is fine for now and we can leave (b) as a future possibility if users are running into such issues frequently.